File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.cc
#include <iostream>

using namespace std;

const int ROOT = 1;
const int MX = 10000 * 100 + 2;
int unused = 2;
bool ch[MX];
int nxt[MX][12];

int c2i(char c){
    return c - '0';
}
void insert(string& s){
    int cur = ROOT;
    for(char c : s){
        if(nxt[cur][c2i(c)] == -1) 
            nxt[cur][c2i(c)] = unused++;
        cur = nxt[cur][c2i(c)];
    }
    ch[cur] = true;
}

bool startsWith(string& s){
    int cur = ROOT;
    for(char c : s){
        if(nxt[cur][c2i(c)] == -1)
            return false;
        cur = nxt[cur][c2i(c)];
        if(ch[cur]) {
          return true;
        }

    }    
    return true;
}

int main()
{
    int t;
    cin >> t;
    while(t--){
        int n;
        cin >> n;
    
        for(int i = 0; i < MX; i++){
            ch[i] = false;
            fill(nxt[i], nxt[i] + 10, -1);
            unused = 2;
        }
        
        bool flag = true;
        for(int i = 0; i < n; i++){
            string v;
            cin >> v;
            if(startsWith(v)){
                 flag = false;
            }
            insert(v);
        }
        
        if(flag) cout << "YES\n";
        else cout << "NO\n";
    }
    
    return 0;
}

File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int ROOT = 1;
    static int MX = 10000 * 10 + 5;
    static int nxt[][] = new int[MX][26];
    static boolean chk[] = new boolean[MX];
    static int idx = 2;

    static void insert(String s) {
        int cur = ROOT;
        for (char c : s.toCharArray()) {
            if (nxt[cur][c2i(c)] == -1) {
                nxt[cur][c2i(c)] = idx++;
            }
            cur = nxt[cur][c2i(c)];
        }
        chk[cur] = true;
    }
    // 길이순 정렬 -> 같을 땐, 숫자 낮은게 앞에 오게

    static boolean find(String s) {
        int cur = ROOT;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (nxt[cur][c2i(c)] == -1) {
                return false;
            }
            cur = nxt[cur][c2i(c)];
        }
        return true;
    }

    static int c2i(char c) {
        return c - '0';
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();
        for (int t = 1; t <= T; t++) {

            for (int i = 0; i < MX; i++) {
                chk[i] = false;
                for (int j = 0; j < 26; j++) {
                    nxt[i][j] = -1;
                }
            }
            idx = 2;

            int N = Integer.parseInt(br.readLine());
            String arr[] = new String[N];
            for (int i = 0; i < N; i++) {
                arr[i] = br.readLine();
            }

            Arrays.sort(arr, (o1, o2) -> {
                if (o1.length() == o2.length()) {
                    return o2.compareTo(o1);
                }

                return o2.length() - o1.length();
            });

            String ret = "YES";
            for (int i = 0; i < N; i++) {

                if (find(arr[i])) {
                    ret = "NO";
                }
                insert(arr[i]);
            }

            sb.append(ret).append("\n");
        }

        System.out.println(sb);
    }
}